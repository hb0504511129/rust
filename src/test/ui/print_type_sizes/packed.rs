// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

// compile-flags: -Z print-type-sizes

// This file illustrates how packing is handled; it should cause
// the elimination of padding that would normally be introduced
// to satisfy alignment desirata.

#![feature(untagged_unions)]

#![allow(dead_code)]

#[derive(Default)]
#[repr(packed)]
struct Packed {
    a: bool,
    b: bool,
    g: i32,
    c: bool,
    w: i64,
    d: bool,
}

#[derive(Default)]
struct Padded {
    a: bool,
    b: bool,
    g: i32,
    c: bool,
    w: i64,
    d: bool,
}

pub fn main() {
    let _c: Packed = Default::default();
    let _d: Padded = Default::default();
}
