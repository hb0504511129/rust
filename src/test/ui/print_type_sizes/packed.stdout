print-type-size type: `Padded`: 32 bytes, alignment: 8 bytes
print-type-size     field `.a`: 1 bytes
print-type-size     field `.b`: 1 bytes
print-type-size     padding bytes: 2
print-type-size     field `.g`: 4 bytes, alignment: 4 bytes
print-type-size     field `.c`: 1 bytes
print-type-size     padding bytes: 7
print-type-size     field `.w`: 8 bytes, alignment: 8 bytes
print-type-size     field `.d`: 1 bytes
print-type-size     end padding bytes: 7
print-type-size type: `Packed`: 16 bytes, alignment: 1 bytes
print-type-size     field `.a`: 1 bytes
print-type-size     field `.b`: 1 bytes
print-type-size     field `.g`: 4 bytes
print-type-size     field `.c`: 1 bytes
print-type-size     field `.w`: 8 bytes
print-type-size     field `.d`: 1 bytes
